### DISCUSSION TITLE
Consulta de NFE

### DISCUSSION THREAD

**Discussion by enieber** [BrasilAPI/BrasilAPI#214] [Ideas]
Existe um projeto muito bacana que seria legal integralo ao BrasilAPI, já coloquei uma issue no projeto falando da ideia, caso alguém queria/possa implementar [segue o link da issue](https://github.com/ymoreiratiti/consultaNFE/issues/8)

**Comment by guilhermomg** [comment]
@enieber tava pensando mesmo nesses serviços de emissão/consulta de notas! Sem dúvida são dos mais importantes por aí. 
Penso bastante nos serviços de emissão de nota de serviço, são por prefeitura, mas a maioria das emissões mais estáveis são serviços pagos. Já implementei tanto NF-e quanto NFS-e em C# então estou disposto a ajudar nesse projeto aqui também!

**Comment by enieber** [comment]
@guilhermomg se já quiser dar uma olhada no projeto e implementar a consulta de outros estados fique avontade, caso precise de alguma ajuda com js pode me chamar, não entendo muito de NFS-E ou NFE mas posso ajudar com js

**Comment by lemonte** [comment]
@guilhermomg e @enieber  achei muito bacana a idéia caso precise de alguma ajuda com js pode entrar em contato também! não tenho muita experiência com NF-e ou NFS mas em relação ao JS eu posso dar uma ajuda a vocês também!

**Comment by natanael127** [comment]
Pensei nisso, mas não só com relação às NFE's mas também aos cupons fiscais que têm um QR code com a chave do cupom, CPF, valor da compra, etc...
Esses cupons em geral são estaduais e cada estado deve ter suas regras.

**Comment by lemonte** [comment]
Tem um pacote bem bacana que vale a pena dar uma olhada.
Mas ele so funciona para alguns estados ainda 
https://www.npmjs.com/package/nfe-biblioteca

ELe segue a ideia do QR code e você consegue pegar diversas informações bem bacanas

**Comment by adrianteixeira** [comment]
> Pensei nisso, mas não só com relação às NFE's mas também aos cupons fiscais que têm um QR code com a chave do cupom, CPF, valor da compra, etc...
> Esses cupons em geral são estaduais e cada estado deve ter suas regras.

Acho que esse seria um ganho absurdo, todo dia é um dev a mais passando pelos problemas de consulta de NFE. Caso se inicie, estarei contribuindo com consultas de NFE's no Rio de Janeiro.

**Comment by alexandresanlim** [comment]
O problema dessa implementação é que ele faz um "Crawler", se vcs olharem o vídeo do Teló "Por que eu virei um problema..." ele fala sobre a partir do minuto 1:35, está implementação é extremamente frágil pois depende da não mudança do layout da pagina.  😔

**Comment by enieber** [comment]
esse é um bom ponto @alexandresanlim

**Comment by enieber** [comment]
dado que n temos uma api para consulta, ao menos não conheço uma, teriamos que fazer um velho crawler mesmo..

**Comment by adrianteixeira** [comment]
Eu acho válido partir do princípio do Crawler, não podemos esperar uma api ou algo pronto.. Esse problema é antigo e as soluções sempre param no inicio pq ninguém curte o crawler(com razão) rs

**Comment by alexandresanlim** [comment]
@enieber  pois é... incrível como o governo tem dificuldade de criar APIs ;(

**Comment by enieber** [comment]
achei um otimo ponto de partida para fazer os outros crawler:

https://github.com/nfephp-org/sped-nfe/blob/48084875449509ab64d21679ef135f5dd3413bc8/storage/uri_consulta_nfce.json

**Comment by ymoreiratiti** [comment]
Acredito que não seja difícil de incorporar o [consultaNFE](https://github.com/ymoreiratiti/consultaNFE) neste projeto. 

Seguindo o exemplo do [CEP](https://github.com/BrasilAPI/BrasilAPI/blob/master/pages/api/cep/v1/%5Bcep%5D.js), ele basicamente chama uma biblioteca que esta pronta e já existe (cep-promise) e trata como se fosse um "proxy".

Acredito que podemos manter a biblioteca (consultaNFE) na estrutura atual para quem quiser importar diretamente em seu repositório, assim como inserir a mesma aqui para quem quiser chamar através de uma aplicação mobile/front-end.

**Comment by ymoreiratiti** [comment]
> achei um otimo ponto de partida para fazer os outros crawler:
> 
> https://github.com/nfephp-org/sped-nfe/blob/48084875449509ab64d21679ef135f5dd3413bc8/storage/uri_consulta_nfce.json

Podemos usar esse endereço como base para montar os crawlers. 
O que precisamos nesse caso é basicamente algumas notas para montar os casos de uso.

**Comment by adrianteixeira** [comment]
Pessoal, estou com dificuldade aqui em reproduzir oque o browser faz em requisicoes http.

`http://www4.fazenda.rj.gov.br/consultaNFCe/QRCode?p=33200710834199000107651260002200661002204684|2|1|1|f200be89104d9b1e8aebb08c3b379bc7d3005ce9`

Como exemplo dessa NFCe, parece que ela precisa de um navegador pra executar pois usa script pra carregar um frame dentro da pagina. Alguem passando por isso em outros estados?

**Reply by outlookcast** [reply]
Estou precisando de acessar a mesma informação e estou com o mesmo problema. Conseguiu alguma solução?

**Reply by adrianteixeira** [reply]
@outlookcast já resolvi e implementei num fork a versão para o RJ já consultando corretamente. Segue o link:

https://github.com/adrianteixeira/consultaNFE/blob/master/src/consulta/rj.ts

**Comment by RodriAndreotti** [comment]
Cara, fiz um rastreamento dos requests pelo developer console aqui, e parece que o site obtém algumas informações  sobre esse link e faz um novo post para outra página dentro da infra deles, aqui a URL final (que é reescrita na barra de navegação, inclusive):
http://www4.fazenda.rj.gov.br/consultaNFCe/paginas/resultadoQRCode.faces?cid=3

Mas aqui para mim não carregou nenhuma nota, ficou uma tela em branco.
Inspecionando o código fonte notei que eles usam Javax Faces como render, então pode ser uma pista de como fazer o scrap por ai.

Outro detalhe que notei no developer console, eles retornam todo o HTML dentro de um xml (?!?)
![image](https://user-images.githubusercontent.com/9449048/102068583-5cdc3680-3ddb-11eb-90dc-dff2deaa03e1.png)

![image](https://user-images.githubusercontent.com/9449048/102068649-78dfd800-3ddb-11eb-922d-4bd2764c94e8.png)

Não sei se isso te ajuda, ou se você já tinha achado estas informações também... rs

**Comment by matheuscas** [comment]
Por experiência própria, pessoal, a maioria dos sites, para NFE ou NFCe, o caminho é o crawler. Eu fiz um para um projeto pessoal para a Bahia: https://github.com/matheuscas/pynfce. E teve de ser com crawler pois os sites são bem horríveis.

**Comment by enieber** [comment]
acho q o maior problema é sites que fazem o bloqueio de crowlers
